tests/unit-tests/util/class-wc-tests-core-functions.php

```diff
@@ -694,4 +725,101 @@ public function test_wc_selected() {
 			$this->assertEquals( wc_selected( $value, $options ), $actual_result );
 		}
 	}
+
+	/**
+	 * Test wc_array_cartesian.
+	 *
+	 * @return void
+	 */
+	public function test_wc_array_cartesian() {
+		$product    = new WC_Product_Variable();
+		$attributes = array();
+
+		$attribute = new \WC_Product_Attribute();
+		$attribute->set_id( 0 );
+		$attribute->set_name( 'Attr1' );
+		$attribute->set_options( array( 'Attr1 Value 1', 'Attr1 Value 2' ) );
+		$attribute->set_position( 0 );
+		$attribute->set_visible( true );
+		$attribute->set_variation( true );
+		$attributes[] = $attribute;
+
+		$attribute = new \WC_Product_Attribute();
+		$attribute->set_id( 0 );
+		$attribute->set_name( 'Attr2' );
+		$attribute->set_options( array( 'Attr2 Value 1', 'Attr2 Value 2' ) );
+		$attribute->set_position( 0 );
+		$attribute->set_visible( true );
+		$attribute->set_variation( true );
+		$attributes[] = $attribute;
+
+		$product->set_props( array(
+			'attributes' => $attributes,
+		) );
+		$product->save();
+
+		$variation_attributes  = wc_list_pluck( array_filter( $product->get_attributes(), 'wc_attributes_array_filter_variation' ), 'get_slugs' );
+		$expected_combinations = array(
+			array(
+				'attr2' => 'Attr2 Value 1',
+				'attr1' => 'Attr1 Value 1',
+			),
+			array(
+				'attr2' => 'Attr2 Value 2',
+				'attr1' => 'Attr1 Value 1',
+			),
+			array(
+				'attr2' => 'Attr2 Value 1',
+				'attr1' => 'Attr1 Value 2',
+			),
+			array(
+				'attr2' => 'Attr2 Value 2',
+				'attr1' => 'Attr1 Value 2',
+			),
+		);
+
+		$this->assertEquals( $expected_combinations, array_reverse( wc_array_cartesian( $variation_attributes ) ) );
```

Comment:
This test [seems to fail sometimes](https://travis-ci.org/woocommerce/woocommerce/builds/402659882?utm_source=github_status&utm_medium=notification). There may be a bug in `wc_array_cartesian` where the order isn't the same every time or (more likely) `$variation_attributes` needs to be sorted before passing into `wc_array_cartesian`.

Reply:
If this is a unit-test for function `wc_array_cartesian`, then I think we should pass two predefined arrays to the function and check the result and not make it depend on results of another, unrelated function.

But it's nice that these irregularities surface when writing tests..